### ISSUE TITLE
DepthToSpace produce GPU address fault with Metal delegate

### ISSUE THREAD

**Issue by gyujin-kim** [tensorflow/tensorflow#60158]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Others

### Have you reproduced the bug with TF nightly?

Yes

### Source

binary

### Tensorflow Version

2.11.0

### Custom Code

No

### OS Platform and Distribution

macOS Ventura 13.2.1

### Mobile device

iPhone12

### Current Behaviour?

```shell
I am testing 2x and 4x super-resolution using tflite on iphone using metal delegate.
The two models use almost the same network and determine the multiplier in DepthToSpace,
but 'GPU address fault' occurs at 4x.

--Execution of the command buffer was aborted due to an error during execution. Caused GPU Address Fault Error (0000000b:kIOGPUCommandBufferCallbackErrorPageFault)

When I checked the detailed log through the shader validation option, 'bad memory access (over access)' was occurring.

-- buffer: <unnamed>, length:4915200, resident:Read Write
-- [GPUDebug] Invalid device load executing kernel function "ComputeFunction" encoder: "", dispatch: 0, at offset 4915232
-- file:///program_source:35:0 - ComputeFunction()

The peculiarity is that the fault address offset is not constant.
(ex: 4915248, 4917808, 4918176, 4917792, 4917776, 4918240, 4915216, ...)

Here is the tflite I used
https://drive.google.com/drive/folders/1LAQjbJeXaeGQrhNc-fDiFRfsIyO5tGBR?usp=share_link
```

### Standalone code to reproduce the issue

```shell
It is 1-channel(luminance) super-resolution(2x, 4x) model with 320x240 resolution.
2x is working well, and 4x is not working. I/O type is float32
```

### Relevant log output

```shell
* with Metal API Validation option
2023-03-28 10:45:30.501062+0900 EditorDemo[2735:416512] TensorFlow Lite version : 2.11.0
2023-03-28 10:45:30.503471+0900 EditorDemo[2735:416512] Created TensorFlow Lite delegate for Metal.
INFO: Created TensorFlow Lite delegate for Metal.
2023-03-28 10:45:30.513366+0900 EditorDemo[2735:416512] Initialized TensorFlow Lite runtime.
INFO: Initialized TensorFlow Lite runtime.

2023-03-28 10:45:33.801156+0900 EditorDemo[2735:416822] Execution of the command buffer was aborted due to an error during execution. Caused GPU Address Fault Error (0000000b:kIOGPUCommandBufferCallbackErrorPageFault)

2023-03-28 10:45:33.802032+0900 EditorDemo[2735:416822] Execution of the command buffer was aborted due to an error during execution. Caused GPU Address Fault Error (0000000b:kIOGPUCommandBufferCallbackErrorPageFault)

* with Metal Shader Validation option
2023-03-28 10:49:39.145090+0900 EditorDemo[2747:418356] TensorFlow Lite version : 2.11.0
2023-03-28 10:49:39.145399+0900 EditorDemo[2747:418356] Created TensorFlow Lite delegate for Metal.
INFO: Created TensorFlow Lite delegate for Metal.
2023-03-28 10:49:39.146572+0900 EditorDemo[2747:418356] Initialized TensorFlow Lite runtime.
INFO: Initialized TensorFlow Lite runtime.

2023-03-28 10:49:45.131812+0900 EditorDemo[2747:418729] [GPUDebug] Invalid device load executing kernel function "ComputeFunction" encoder: "", dispatch: 0, at offset 4915232
file:///program_source:35:0 - ComputeFunction()
buffer: <unnamed>, length:4915200, resident:Read Write

2023-03-28 10:49:45.134406+0900 EditorDemo[2747:418729] [GPUDebug] Invalid device load executing kernel function "ComputeFunction" encoder: "", dispatch: 0, at offset 4915232
file:///program_source:35:0 - ComputeFunction()
buffer: <unnamed>, length:4915200, resident:Read Write
```
</details>

**Comment by gyujin-kim** [comment]
* Additional Info.
I am using TensorFlow Lite C API with podspec dependency 'TensorFlowLiteC', 'TensorFlowLiteC/Metal' (version is 2.11.0)
The interpreter option is default and delegate option is as followed.
  - allow_precision_loss = true
  - wait_type = TFLGpuDelegateWaitTypePassive
  - enable_quantization = true
 tflite with other target_spec.supported_types(fp16, fp32) has same problem.
(current is default convert option, it has int8 type inside)
And when I running on Android device with same tflite file, it is working well.

**Comment by pjpratik** [comment]
Hi @gyujin-kim Thanks for reporting this issue.

Can you please provide the steps you have followed in order to reproduce the issue?

Are you using the [ESRGAN model](https://tfhub.dev/captain-pool/lite-model/esrgan-tf2/1) from TF Hub? 

Thanks.

**Comment by gyujin-kim** [comment]
Hi @pjpratik 
I am using [ECBSR](https://github.com/xindongzhang/ECBSR).
As mentioned above, i'am using TensorFow C API(not C++).
Current test sequence is as followed.
 1. Input image is RGB and it is resized to model input size & converted to YUV420. 
 2. Only Y-channel data is served to model input as float.
     ( just cast to float, (float)input )
I think, pre-process step can be skipped.
Because this issue is the image offset(shape) problem, 1 channel data of input(ex. R channel data) is served to model, same issue will be produced.
And i can't share current test code cause the company policy.
I will try to attach SR model to another sample app.

Thanks.

**Comment by pjpratik** [comment]
@gyujin-kim Thanks for the information.

@sachinprasadhs Could you please look into this issue. Thanks.

**Comment by gyujin-kim** [comment]
Hi @sachinprasadhs @impjdi 
I applied super-resolution to [tensorflow lite segmentation example](https://github.com/tensorflow/examples/tree/master/lite/examples/image_segmentation/ios).
(https://drive.google.com/file/d/1PoGE9gd1X8HODtl5TvNBQc5aQQWsTEjE/view?usp=share_link)
Although it is different from the environment I actually use (C API vs Swift API), the test result using random input produced the same issue.

Create an interpreter: ImageSegmentationHelper.swift line 64-81
invoke : ImageSegmentationHelper.swift line 161~211
You can test 2x and 4x by modifying ImageSegmentationHelper.swift line 367.
2x works fine, 4x is giving 'GPU address fault'.

Please review.
thank you

**Comment by roserg** [comment]
https://github.com/tensorflow/tensorflow/commit/c6250a0571668e920989f9506c2fbdfc53baff6d
Depth to space fixed in this commit

**Comment by gyujin-kim** [comment]
@roserg 
I confirmed that the problem was resolved in the version 'TensorFlowLiteC 0.0.1-nightly.20230406'.
thank you for your hard work.

Also, thank you for all the members who cared about this issue(@tiruk007, @pjpratik , @sachinprasadhs, @impjdi ).
Thanks.

**Comment by sachinprasadhs** [comment]
@gyujin-kim , Thanks for confirming, could you please close the issue as well, Thanks!
